source/extensions/http/header_validators/envoy_default/http1_header_validator.cc

```diff
@@ -0,0 +1,362 @@
+
+/*
+ * Header validation implementation for the Http/1 codec. This class follows guidance from
+ * several RFCS:
+ *
+ * RFC 3986 <https://datatracker.ietf.org/doc/html/rfc3986> URI Generic Syntax
+ * RFC 7230 <https://datatracker.ietf.org/doc/html/rfc7230> HTTP/1.1 Message Syntax
+ * RFC 7231 <https://datatracker.ietf.org/doc/html/rfc7231> HTTP/1.1 Semantics and Content
+ *
+ */
+Http1HeaderValidator::Http1HeaderValidator(const HeaderValidatorConfig& config, Protocol protocol,
+                                           StreamInfo::StreamInfo& stream_info)
+    : HeaderValidator(config, protocol, stream_info) {}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http1HeaderValidator::validateRequestHeaderEntry(const HeaderString& key,
+                                                 const HeaderString& value) {
+  static const absl::node_hash_map<absl::string_view, HeaderValidatorFunction> kHeaderValidatorMap{
+      {":method", &Http1HeaderValidator::validateMethodHeader},
+      {":authority", &Http1HeaderValidator::validateHostHeader},
+      {"host", &Http1HeaderValidator::validateHostHeader},
+      {":scheme", &Http1HeaderValidator::validateSchemeHeader},
+      {":path", &Http1HeaderValidator::validateGenericPathHeader},
+      {"transfer-encoding", &Http1HeaderValidator::validateTransferEncodingHeader},
+      {"content-length", &Http1HeaderValidator::validateContentLengthHeader},
+  };
+
+  const auto& key_string_view = key.getStringView();
+  if (key_string_view.empty()) {
+    // reject empty header names
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().EmptyHeaderName};
+  }
+
+  auto validator_it = kHeaderValidatorMap.find(key_string_view);
+  if (validator_it != kHeaderValidatorMap.end()) {
+    const auto& validator = validator_it->second;
+    return (*this.*validator)(value);
+  }
+
+  if (key_string_view.at(0) != ':') {
+    // Validate the (non-pseudo) header name
+    auto name_result = validateGenericHeaderName(key);
+    if (!name_result) {
+      return name_result;
+    }
+  }
+
+  return validateGenericHeaderValue(value);
+}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http1HeaderValidator::validateResponseHeaderEntry(const HeaderString& key,
+                                                  const HeaderString& value) {
+  const auto& key_string_view = key.getStringView();
+  if (key_string_view.empty()) {
+    // reject empty header names
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().EmptyHeaderName};
+  }
+
+  if (key_string_view == ":status") {
+    // Validate the :status header against the RFC valid range (100 <= status <= 599)
+    return validateStatusHeader(StatusPseudoHeaderValidationMode::ValueRange, value);
+  } else if (key_string_view == "content-length") {
+    // Validate the Content-Length header
+    return validateContentLengthHeader(value);
+  } else if (key_string_view.at(0) != ':') {
```

Comment:
Do we need to check if the name does not start with `:`? The only pseudo header that Envoy synthesizes for H/1 responses is `:status` so everything else that starts with `:` should be rejected.